Creates protected route

GurkirtSingh opened the pull request:
### What this PR does (required):
- Adds new component to protect route
- Protects existing route that should be access by authenticated user

### Any information needed to test this feature (required):
- Visit `/dashboard` route
- Route should only be access if user is logged in

### Any issues with the current functionality (optional):
- If user try to access a protected page and is not logged in, page will redirect to `/dashboard` after login/signup no matter which page you try to access before coming to login/signup page

GurkirtSingh committed:
created 'protected route' function component

GurkirtSingh committed:
protected existing route

Using new  'protected route' component protected existing route that should be access by authenticated user

rajivtitus commented on client/src/components/ProtectedRoute/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,51 @@
+  const { loggedInUser } = useAuth();
+  const { initSocket } = useSocket();
+
+  useEffect(() => {
```
You can remove this `useEffect` as we don't need to initialize the socket connection here. Ideally, we would want to do it in the `useSocketContext`

GurkirtSingh replied:
Okay, I'll remove it

rajivtitus commented on client/src/components/ProtectedRoute/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,51 @@
+        if (loggedInUser) {
+          // render protected component on valid user
+          if (component) {
+            return React.createElement(component, props);
```
We would use React.createElement() instead of JSX when we do not want to set up compilation for our project, I don't think we need to use that approach here though :)

GurkirtSingh replied:
Okay, I'll go with required component approach with component being any type and rendering it.

rajivtitus commented on client/src/components/ProtectedRoute/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  return (
+    <Route
```
```suggestion
    return <Route {... rest} render={(props) => (loggedInUser ? <Component {...props} /> : <Redirect to="/login" />)} />;
```

We can simplify this component further!

rajivtitus commented on client/src/components/ProtectedRoute/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,51 @@
+import CircularProgress from '@mui/material/CircularProgress';
+import { Route, Redirect, RouteProps, RouteComponentProps } from 'react-router-dom';
+
+const ProtectedRoute: React.FC<RouteProps> = ({ component, render, ...routeProps }) => {
```
```suggestion
const ProtectedRoute: React.FC<RouteProps> = ({ component: Component, ...rest }) => {
```

rajivtitus replied:
```suggestion
const ProtectedRoute = ({ component: Component, ...rest }: RouteProps) : JSX.Element  => {
```

Short list of React.FC cons:

- Provides an implicit definition of children, even if your component doesn't need to have children. That might cause an error.
- Doesn't support generics.
- Doesn't work correctly with defaultProps.

Additional Reference: https://github.com/facebook/create-react-app/pull/8177

rajivtitus reviewed (commented):
Good work with this! Just some comments for you to have a look and address 👍🏽

GurkirtSingh committed:
Update client/src/components/ProtectedRoute/ProtectedRoute.tsx

Co-authored-by: Rajiv Titus <75184965+rajivtitus@users.noreply.github.com>

GurkirtSingh committed:
Update client/src/components/ProtectedRoute/ProtectedRoute.tsx

Co-authored-by: Rajiv Titus <75184965+rajivtitus@users.noreply.github.com>

GurkirtSingh committed:
Fixed: Protected Route Component

Extended Route Props to change component filed from optional to required
Removed Socket updates.
Simplified return statement.

rajivtitus reviewed (approved):
LGTM! Thanks for making those changes 👍🏽
